crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -123,46 +100,95 @@ impl ChainId {
     /// assert_eq!(ChainId::is_epoch_format("chainA"), false);
     /// assert_eq!(ChainId::is_epoch_format("chainA-1"), true);
     /// assert_eq!(ChainId::is_epoch_format("c-1"), true);
+    /// assert_eq!(ChainId::is_epoch_format("-1"), false);
     /// ```
     pub fn is_epoch_format(chain_id: &str) -> bool {
-        let re = safe_regex::regex!(br".*[^-]-[1-9][0-9]*");
-        re.is_match(chain_id.as_bytes())
+        Self::split_chain_id(chain_id).1.is_some()
     }
 
     /// with_version() checks if a chain_id is in the format required for parsing epochs, and if so
     /// replaces it's version with the specified version
     /// ```
     /// use ibc::core::ics24_host::identifier::ChainId;
-    /// assert_eq!(ChainId::new("chainA".to_string(), 1).with_version(2), ChainId::new("chainA".to_string(), 2));
+    /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
     pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+        if self.version != 0 {
+            if let (name, Some(_)) = Self::split_chain_id(&self.id) {
+                self.id = format!("{name}-{version}");
+                self.version = version;
+            }
         }
         self
     }
+
+    /// Splits chain_id into name and version if the id includes epoch number.
+    ///
+    /// Chain id with epoch number has format `{name}-{version}` where version
+    /// is a positive integer starting with non-zero digit.  If `chain_id` is in
+    /// that format, returns `(name, Some(version))`; otherwise returns
+    /// `(chain_id, None)`.
+    ///
+    /// **Panics** if chain id is in epoch format but the version overflows
+    /// `u64`.
+    fn split_chain_id(chain_id: &str) -> (&str, Option<u64>) {
+        fn split(chain_id: &str) -> Option<(&str, u64)> {
+            let (name, version) = chain_id.rsplit_once('-')?;
+            if name.is_empty() {
+                return None;
+            }
+            let (car, cdr) = version.as_bytes().split_first()?;
+            if b'1' <= *car && *car <= b'9' && cdr.iter().all(u8::is_ascii_digit) {
+                Some((name, version.parse().unwrap()))
```

Comment:
As mentioned in other comments, we'd need to return an `IdentifierError` when the version parsing here fails